packages/keyboard-shortcuts/src/context.js

```diff
@@ -3,4 +3,13 @@
  */
 import { createContext } from '@wordpress/element';
 
-export const context = createContext();
+const globalShortcuts = { current: new Set() };
+document.addEventListener( 'DOMContentLoaded', () => {
+	document.body.addEventListener( 'keydown', ( event ) => {
+		for ( const keyboardShortcut of globalShortcuts.current ) {
+			keyboardShortcut( event );
+		}
+	} );
+} );
```

Comment:
> there's no effect if you don't have any provider

But that only means that there is a good design reason to have a provider. And it doesn't need to be exactly a provider, the existing `KeyboardShortcuts.Register` component is a perfect fit for this listener-registering effect. It already has an effect that registers stuff in the data store.

Reply:
`KeyboardShortcuts.Register` is a "user" API not something that is provided by the library `keyboard-shortcuts`.

The only thing the library provides now are "registerShortcut" actions and "useShortcut" hooks. If you mean `useShortcut` should trigger the "global event listener" if it's not already listening, that's basically the same as overriding the "add" function to do that.

Reply:
> If you mean `useShortcut` should trigger the "global event listener" if it's not already listening

Yes, that should work. My request is to make the listener controlled by React code. It should be added when a certain React component or hook is mounted, and automatically removed when it's unmounted. A reference-counted global variable and an effect can make it work:
```js
let listenerCounter = 0;

function useShortcut() {
  useEffect( () => {
    if ( listenerCounter++ === 0 ) {
      document.body.addListener();
    }
    return () => {
      if ( --listenerCounter === 0 ) {
        document.body.removeListener();
      }
    };
  }, [] );
  /* rest of useShortcut code */
}
```

Reply:
@jsnajdr Why a lingering global event listener is a problem for you?

Reply:
The global listener currently has two practical issues: the static module-level access to `document` will break in SSR environments, and the registration on `DOMContentLoaded` event makes some assumptions that don't need to be true.

Code that's triggered by merely importing a module, as a side-effect, without ever using any of its exports, has a propensity to cause trouble for the module users. If you disagree, you can keep the global listener, but I think that at least the two issues mentioned should be fixed.

Reply:
No, I agree with these two downsides, what I'm wondering is why not just hook the event on "add" in the "context" instead of moving this responsibility to a single useShortcut call which is not global right.

Reply:
> why not just hook the event on "add" in the "context"

You mean extending the `Set.prototype.add` method that is called in `useShortcut`, inside `useEffect`, as `shortcuts.current.add()`? Yes, that's fine, because it's exactly the same thing as I proposed above, only slightly reshuffled. Or isn't it? To detect the "first `add`" and "last `delete`", you'll also need the reference counter. I think we are in a complete agreement 🙂

Reply:
Updated